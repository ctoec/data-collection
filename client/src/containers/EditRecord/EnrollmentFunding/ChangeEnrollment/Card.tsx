import React, { useState, useEffect } from 'react';
import {
  Card,
  ExpandCard,
  Button,
  CardExpansion,
  Alert,
  CardProps,
} from '@ctoec/component-library';
import { Enrollment, Child } from '../../../../shared/models';
import { ChangeEnrollmentForm } from './Form';

type ChangeEnrollmentCardProps = CardProps & {
  child: Child;
  currentEnrollment?: Enrollment;
  key?: string;
  activeKey?: string;
  afterSaveSuccess: () => void;
};

/**
 * Component for gathering user input to change child's Enrollment.
 * Uses a ChangeEnrollment data object to enable the user to provide
 * enrollment end date and funding last reporting period for previously current data.
 */
export const ChangeEnrollmentCard: React.FC<ChangeEnrollmentCardProps> = ({
  child,
  currentEnrollment,
  afterSaveSuccess,
  expanded,
}) => {
  const [expandedCard, setExpandedCard] = useState(expanded);
  const [closeCard, setCloseCard] = useState(false);
  const [error, setError] = useState<string>();

  // Explicitly don't want `closeCard` as a dep, as this
  // needs to be triggered on render caused by child refetch
  // (not only when closeCard changes)
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    if (closeCard) setCloseCard(false);
  });

  return (
    <Card forceClose={closeCard}>
      <div className="display-flex flex-justify flex-row flex-align-center">
        {!currentEnrollment ? (
          <div className="usa-prose-body">
            {child.firstName} has no active enrollments
          </div>
        ) : (
          <div className="usa-prose-body">
            Has {child.firstName}'s age group and/or site changed?
          </div>
        )}
        <ExpandCard>
          <Button text="Change enrollment" appearance="outline" />
        </ExpandCard>
      </div>
      <CardExpansion>
        <h3 className="margin-top-2 margin-bottom-2">New enrollment</h3>
        {error && <Alert type="error" text={error} />}
        <ChangeEnrollmentForm
          afterSaveSuccess={() => {
            setCloseCard(true);
            afterSaveSuccess();
          }}
          afterSaveFailure={(err) => setError(err)}
          child={child}
          currentEnrollment={currentEnrollment}
        />
      </CardExpansion>
    </Card>
  );
};
