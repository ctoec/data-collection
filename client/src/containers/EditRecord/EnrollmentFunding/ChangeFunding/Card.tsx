import React, { useState } from 'react';
import { Card, Button } from '@ctoec/component-library';
import { Enrollment } from '../../../../shared/models';
import { ChangeFundingForm } from './Form';
import { RecordFormProps } from '../../../../components/Forms';
import { Heading, HeadingLevel } from '../../../../components/Heading';
import { getCurrentFunding } from '../../../../utils/models';

type ChangeFundingCardProps = {
  enrollment: Enrollment;
  orgId: number;
  afterSaveSuccess: () => void;
  setAlerts: RecordFormProps['setAlerts'];
  topHeadingLevel: HeadingLevel;
};

/**
 * Component for gathering user input to change enrollment's funding.
 * Uses a ChangeFunding data object to enable the user to provide
 * new funding data and/or last reporting period for previously current
 * funding, based on user's chosen action (start new funding vs end current funding)
 */
export const ChangeFundingCard: React.FC<ChangeFundingCardProps> = ({
  enrollment,
  orgId,
  afterSaveSuccess,
  setAlerts,
  topHeadingLevel,
}) => {
  const [visibleForm, setVisibleForm] = useState<'end' | 'start'>();
  const activeFunding = getCurrentFunding({ enrollment })
	
	return (
    <>
      {!!visibleForm && (
        <Card>
          <>
            <Heading level={topHeadingLevel}>
              {visibleForm === 'end' ? 'End current funding' : 'Change funding'}
            </Heading>
            <ChangeFundingForm
              afterSaveSuccess={() => {
                setVisibleForm(undefined);
                afterSaveSuccess();
              }}
              afterSaveFailure={(err) => {
                console.error(err);
                setAlerts([
                  {
                    type: 'error',
                    text:
                      'Unable to change funding. Make sure all required information is provided',
                  },
                ]);
              }}
              changeType={visibleForm}
              enrollment={enrollment}
              hideForm={() => setVisibleForm(undefined)}
              orgId={orgId}
            />
          </>
        </Card>
      )}
      <Card>
        <div className="display-flex flex-column flex-align-end">
          <div>
            {activeFunding && (
              <Button
                className="margin-right-4"
                text="End current funding"
                appearance="unstyled"
                onClick={() => setVisibleForm('end')}
              />
            )}
            <Button
              text={activeFunding ? 'Change funding' : 'Start new funding'}
              appearance="unstyled"
              onClick={() => setVisibleForm('start')}
            />
          </div>
        </div>
      </Card>
    </>
  );
};
