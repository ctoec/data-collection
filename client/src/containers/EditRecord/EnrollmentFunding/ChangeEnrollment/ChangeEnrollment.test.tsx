import React from 'react';
import {
  snapshotTestHelper,
  accessibilityTestHelper,
} from '../../../../testHelpers';
import { ChangeEnrollmentCard } from './Card';
import moment from 'moment';
import {
  Family,
  Site,
  AgeGroup,
  ReportingPeriod,
  Funding,
  Child,
  FundingSource,
} from '../../../../shared/models';
import { RenderResult } from '@testing-library/react';
import { fireEvent, wait } from '@testing-library/dom';

const enrollment = {
  child: {} as Child,
  id: 1,
  site: { id: 1, siteName: 'Site' } as Site,
  ageGroup: AgeGroup.InfantToddler,
  entry: moment.utc('2020-09-01'),
  fundings: [
    {
      id: 1,
      fundingSpace: {
        ageGroup: AgeGroup.InfantToddler,
        source: FundingSource.CDC,
      },
      firstReportingPeriod: {} as ReportingPeriod,
    } as Funding,
  ],
};

const child = {
  id: '0000000-0000-0000-0000-0000000000000',
  firstName: 'First',
  lastName: 'Toddler',
  birthdate: moment.utc('2019-06-01', 'YYYY-MM-DD'),
  family: {} as Family,
  enrollments: [enrollment],
  organization: { id: 1 },
} as Child;

const waitExpandChangeEnrollment = async (renderResult: RenderResult) => {
  const changeEnrollmentButton = await renderResult.findByText(
    'Change enrollment'
  );
  fireEvent.click(changeEnrollmentButton);
  await wait();
};

describe('EditRecord', () => {
  describe('EnrollmentFunding', () => {
    describe('ChangeEnrollment', () => {
      snapshotTestHelper(
        <ChangeEnrollmentCard
          child={child}
          currentEnrollment={enrollment}
          afterSaveSuccess={jest.fn()}
        />,
        { before: waitExpandChangeEnrollment }
      );

      accessibilityTestHelper(
        <ChangeEnrollmentCard
          child={child}
          currentEnrollment={enrollment}
          afterSaveSuccess={jest.fn()}
        />,
        { before: waitExpandChangeEnrollment }
      );
    });
  });
});
