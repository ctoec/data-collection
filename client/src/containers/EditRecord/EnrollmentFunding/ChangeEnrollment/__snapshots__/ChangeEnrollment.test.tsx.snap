// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditRecord EnrollmentFunding ChangeEnrollment matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="oec-card oec-card--stacked"
  >
    <div
      class="oec-card-cell"
    >
      <div
        class="display-flex flex-justify flex-row flex-align-center"
      >
        <div
          class="usa-prose-body"
        >
          Has First's age group and/or site changed?
        </div>
        <button
          class="usa-button usa-button--outline bx--btn bx--btn--primary"
          tabindex="0"
          type="button"
        >
          Change enrollment
        </button>
      </div>
    </div>
    <div
      class="oec-card-divider"
    />
    <div
      class="oec-card-cell oec-card-expansion"
    >
      <form
        class="usa-form usa-form"
        id="change-enrollment-form"
      >
        <h2
          class="margin-top-2 margin-bottom-2"
        >
          New enrollment
        </h2>
        <fieldset
          aria-labelledby="fieldset-legend-site-radiogroup"
          class="usa-fieldset"
          id="site-radiogroup"
        >
          <legend
            id="fieldset-legend-site-radiogroup"
          >
            <span
              class="usa-label"
            >
              Site
            </span>
          </legend>
          <div />
        </fieldset>
        <fieldset
          aria-labelledby="fieldset-legend-start-date start-date-hint"
          class="usa-fieldset"
          id="start-date"
        >
          <legend
            id="fieldset-legend-start-date"
          >
            <span
              class="usa-label"
            >
              Enrollment start date
            </span>
          </legend>
          <span
            class="usa-hint text-italic"
            id="start-date-hint"
          >
            For example: 06/21/2008
          </span>
          <div>
            <div
              class="bx--form-item"
            >
              <div
                class="bx--date-picker bx--date-picker--single"
                id="start-date-picker"
              >
                <div
                  class="bx--date-picker-container"
                >
                  <label
                    class="bx--label bx--visually-hidden"
                    for="start-date-input"
                  >
                    Date picker
                  </label>
                  <div
                    class="bx--date-picker-input__wrapper"
                  >
                    <input
                      class="bx--date-picker__input flatpickr-input"
                      id="start-date-input"
                      pattern="\\\\d{1,2}\\\\/\\\\d{1,2}\\\\/\\\\d{4}"
                      placeholder="mm/dd/yyyy"
                      type="text"
                      value=""
                    />
                    <svg
                      aria-hidden="true"
                      class="bx--date-picker__icon"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M26,4h-4V2h-2v2h-8V2h-2v2H6C4.9,4,4,4.9,4,6v20c0,1.1,0.9,2,2,2h20c1.1,0,2-0.9,2-2V6C28,4.9,27.1,4,26,4z M26,26H6V12h20	V26z M26,10H6V6h4v2h2V6h8v2h2V6h4V10z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </fieldset>
        <fieldset
          aria-labelledby="fieldset-legend-care-model-radiogroup"
          class="usa-fieldset"
          id="care-model-radiogroup"
        >
          <legend
            id="fieldset-legend-care-model-radiogroup"
          >
            <span
              class="usa-label"
            >
              Care model
            </span>
          </legend>
          <div>
            <span>
              <div
                class="usa-radio"
              >
                <input
                  class="usa-radio__input"
                  id="In-person"
                  name="model"
                  type="radio"
                  value="In-person"
                />
                <label
                  class="usa-radio__label"
                  for="In-person"
                >
                  In-person
                </label>
              </div>
            </span>
            <span>
              <div
                class="usa-radio"
              >
                <input
                  class="usa-radio__input"
                  id="Hybrid"
                  name="model"
                  type="radio"
                  value="Hybrid"
                />
                <label
                  class="usa-radio__label"
                  for="Hybrid"
                >
                  Hybrid
                </label>
              </div>
            </span>
            <span>
              <div
                class="usa-radio"
              >
                <input
                  class="usa-radio__input"
                  id="Distance"
                  name="model"
                  type="radio"
                  value="Distance"
                />
                <label
                  class="usa-radio__label"
                  for="Distance"
                >
                  Distance
                </label>
              </div>
            </span>
            <span>
              <div
                class="usa-radio"
              >
                <input
                  class="usa-radio__input"
                  id="Unknown"
                  name="model"
                  type="radio"
                  value="Unknown"
                />
                <label
                  class="usa-radio__label"
                  for="Unknown"
                >
                  Unknown
                </label>
              </div>
            </span>
          </div>
        </fieldset>
        <fieldset
          aria-labelledby="fieldset-legend-age-group-radiogroup"
          class="usa-fieldset"
          id="age-group-radiogroup"
        >
          <legend
            id="fieldset-legend-age-group-radiogroup"
          >
            <span
              class="usa-label"
            >
              Age group
            </span>
          </legend>
          <div>
            <span>
              <div
                class="usa-radio"
              >
                <input
                  class="usa-radio__input"
                  id="Infant/toddler"
                  name="ageGroup"
                  type="radio"
                  value="Infant/toddler"
                />
                <label
                  class="usa-radio__label"
                  for="Infant/toddler"
                >
                  Infant/toddler
                </label>
              </div>
            </span>
            <span>
              <div
                class="usa-radio"
              >
                <input
                  class="usa-radio__input"
                  id="Preschool"
                  name="ageGroup"
                  type="radio"
                  value="Preschool"
                />
                <label
                  class="usa-radio__label"
                  for="Preschool"
                >
                  Preschool
                </label>
              </div>
            </span>
            <span>
              <div
                class="usa-radio"
              >
                <input
                  class="usa-radio__input"
                  id="School-aged"
                  name="ageGroup"
                  type="radio"
                  value="School aged"
                />
                <label
                  class="usa-radio__label"
                  for="School-aged"
                >
                  School aged
                </label>
              </div>
            </span>
          </div>
        </fieldset>
        <fieldset
          aria-labelledby="fieldset-legend-funding-source-undefined funding-source-undefined-hint"
          class="usa-fieldset"
          id="funding-source-undefined"
        >
          <legend
            id="fieldset-legend-funding-source-undefined"
          >
            <span
              class="usa-label"
            >
              Funding source options
            </span>
          </legend>
          <span
            class="usa-hint text-italic"
            id="funding-source-undefined-hint"
          >
            Choose an age group to see available funding types
          </span>
          <div />
        </fieldset>
        <h2>
          Previous enrollment
        </h2>
        <div
          class="usa-form-group"
        >
          <label
            class="usa-label"
            for="last-reporting-period"
          >
            Last reporting period for current CDC - Child Day Care - undefined funding
          </label>
          <span
            class="usa-hint text-italic"
          >
            Last reporting period is only required if this child's funding is no longer active.
          </span>
          <select
            aria-required="false"
            class="usa-select"
            id="last-reporting-period"
            name="last-reporting-period"
          >
            <option
              value=""
            >
              - Select -
            </option>
          </select>
        </div>
        <button
          class="usa-button usa-button--outline bx--btn bx--btn--primary"
          tabindex="0"
          type="button"
        >
          Cancel
        </button>
        <button
          class="usa-button bx--btn bx--btn--primary"
          tabindex="0"
          type="submit"
        >
          Change enrollment
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
